src/trovares_connector/neo4j_connector/query_translator.py

```diff
@@ -0,0 +1,366 @@
+
+        walker = ParseTreeWalker()
+        xlater = Xlater(verbose = self._verbose)
+        walker.walk(xlater, tree)
+
+        self._rewrites = dict()
+        for edge in xlater.edges:
+            self._add_rewrites_for_edge_frame_names(edge)
+
+        new_query = self._rewrite_query(query)
+        del self._rewrites
+        return new_query
+
+    def _add_rewrites_for_edge_frame_names(self, edge) -> None:
+        if self._verbose:
+            print("In _add_rewrites_for_edge_frame_names")
+            pprint(edge)
+        left_nodes = edge[0]
+        relationship = edge[1]
+        right_nodes = edge[2]
+        if 'rel_types' not in relationship:
+            return
+        for frame_name in relationship['rel_types'].keys():
+            if self._verbose:
+                print(f"Exploring frame name: {frame_name}")
+            if frame_name in self._edge_name_mapping:
+                left_node_labels = left_nodes['labels']
+                if len(left_node_labels) == 0:
+                    source = None
+                elif len(left_node_labels) == 1:
+                    source = list(left_node_labels.keys())[0]
+                else:
+                    raise ValueError(f"Support for multi-label left nodes not yet support: {left_nodes}")
+                right_node_labels = right_nodes['labels']
+                if len(right_node_labels) == 0:
+                    target = None
+                elif len(right_node_labels) == 1:
+                    target = list(right_node_labels.keys())[0]
+                else:
+                    raise ValueError(f"Support for multi-label nodes not yet support: {right_nodes}")
+
+                swap_source_target = relationship['left_arrow'] and not relationship['right_arrow']
+                if swap_source_target:
+                    temp = source
+                    source = target
+                    target = temp
+                
+                if self._verbose:
+                    print(f"----> Source: {source}")
+                    print(f"----> Target: {target}")
+
+                if source is not None and target is not None:
+                    mapping = self._edge_name_mapping[frame_name]
+                    if self._verbose:
+                        print(f"Frame to adapt: (:{source})-[:{frame_name}]->(:{target})")
+                        print(f"  Frame Name {frame_name} needs adapting, {len(mapping)} mapped frames")
+                    for variant in mapping:
+                        if variant['source'] == source and variant['target'] == target:
+                            if self._verbose:
+                                print(f"Use this variant: {variant}")
+                            reltype_loc = relationship['rel_types'][frame_name].location
+                            if reltype_loc['offset'] in self._rewrites:
+                                raise ValueError(f"Multiple rewrites at offset: {reltype_loc['offset']}")
```

Comment:
They may get this error if it ever occurs.  It is hard for me to imagine that it would occur.  It means that there are multiple "rewrites" that need to take place at the exact same character offset within the query; something that is probably ambiguous.